### PR TITLE
Adding validation for duplicated IP's.

### PR DESCRIPTION
**Pull request by radkomateusz**
Adding validation for IP's which don't belong to defined subnet.

### CODE REVIEW

**Comment by zmumi** [comment]
When I try to save lan 10.0.0.0/3 I get 0.0.0.0/3. Updates do no change that.
@mrradzio

File: templates/web/index.html [hunk 1]
```diff
@@ -113,12 +132,193 @@
...
+        var tmpEdges = edges.get();
+        for (var edge in tmpEdges) {
+            usedIPs[tmpEdges[edge].properties.ip] = true;
+        }
+        return usedIPs;
+    }
+
+    function getSubnetIP(subnet) {
+        splitted = subnet.split('/');
+        return binToNum(calcNetBinAddress(ipNumToBinaryStr(ipAddressToNumber(splitted[0])), maskToBinaryStr(splitted[1])) );
```
Comment (zmumi) [inline templates/web/index.html]: do you really have to do it all by yourself?

File: templates/web/index.html [hunk 2]
```diff
@@ -113,12 +132,193 @@
...
+    function binToNum(binaryString) {
+        return parseInt(binaryString, 2);
+    }
+
+    function ipAddressToNumber(ip) {
+        var d = ip.split('.');
+        return ((((((+d[0])*256)+(+d[1]))*256)+(+d[2]))*256)+(+d[3]);
+    }
+
+    function calcNetBinAddress(addressBinStr, netmaskBinStr) {
```
Comment (zmumi) [inline templates/web/index.html]: It is hard to comprehend - there is no js library for that?

File: templates/web/index.html [hunk 3]
```diff
@@ -113,12 +132,193 @@
...
+        var ip=number%256;
+        for (var i=1;i<=3;i++)
+        {
+            number=Math.floor(number/256);
+            ip=number%256+'.'+ip;
+        }
+        return ip; // As string
+    }
+
+    var ip2long = function(ip){
```
Comment (zmumi) [inline templates/web/index.html]: 1. don't undestand the name
2. why this is a variable and not a function?

**Comment by zmumi** [comment]
I believe I couldn't maintain that code - it's way too magic and full of poorly described low-level functions.
If you can't use any library for this just extract IP processing functions to some module with nice interface.

**Comment by radkomateusz** [comment]
As you please: 
Now code uses small js library from: [http://www-personal.umich.edu/~parsec/information/code/ip_calc.js.txt](http://www-personal.umich.edu/~parsec/information/code/ip_calc.js.txt)

You wrote: "When I try to save lan 10.0.0.0/3 I get 0.0.0.0/3. Updates do no change that." 
It is working as design. Based on [ip subnet calculator](http://jodies.de/ipcalc?host=10.168.0.1&mask1=3&mask2=) the net addres is 0.0.0.0/3. 

In situation you provide:  an IP belongs to network with given mask. Application processes that and saves correct network address (without host bits) -> 0.0.0.0. 

When network address contains host bits backend throws exceptions.

At this moment application just ensures that field subnet contains correct network without host bits

File: templates/web/index.html [hunk 4]
```diff
@@ -119,9 +138,105 @@
...
+        var usedIPs = getUsedIPs();
+
+        while (usedIPs[IPv4_intA_to_dotquadA(possibleIP_InSubnet)]) {
+            possibleIP_InSubnet += 1;
+        }
+
+        return IPv4_intA_to_dotquadA(possibleIP_InSubnet);
+    }
+
+    function is_IP_In_Subnet(ip, subnet) {
```
Comment (zmumi) [inline templates/web/index.html]: dzwine nazwy...

File: static/lib/ip_converter.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+* Neither the name of the author or other identifiers used by the author (such as nickname or avatar) may be used to endorse or promote products derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+
+    function IPv4_Address( addressDotQuad, netmaskBits ) {
```
Comment (zmumi) [inline static/lib/ip_converter.js:17]: that lib introduces nasty naming but better than no lib

**Pull request merged**